Vinyl/Turntable.swift

```diff
@@ -35,35 +36,35 @@ public final class Turntable: NSURLSession {
         super.init()
     }
     
-    public convenience init(vinyl: Vinyl, turntableConfiguration: TurntableConfiguration, delegateQueue: NSOperationQueue? = nil) {
+    public convenience init(vinyl: Vinyl, turntableConfiguration: TurntableConfiguration, delegate: NSURLSessionDelegate? = nil, delegateQueue: NSOperationQueue? = nil) {
         
-        self.init(configuration: turntableConfiguration, delegateQueue: delegateQueue)
+        self.init(configuration: turntableConfiguration, delegate: delegate, delegateQueue: delegateQueue)
         player = Turntable.createPlayer(vinyl, configuration: turntableConfiguration)
     }
     
-    public convenience init(cassetteName: String, bundle: NSBundle = testingBundle(), turntableConfiguration: TurntableConfiguration = TurntableConfiguration(), delegateQueue: NSOperationQueue? = nil) {
+    public convenience init(cassetteName: String, bundle: NSBundle = testingBundle(), turntableConfiguration: TurntableConfiguration = TurntableConfiguration(), delegate: NSURLSessionDelegate? = nil, delegateQueue: NSOperationQueue? = nil) {
         
         let vinyl = Vinyl(plastic: Turntable.createCassettePlastic(cassetteName, bundle: bundle))
-        self.init(vinyl: vinyl, turntableConfiguration: turntableConfiguration, delegateQueue: delegateQueue)
+        self.init(vinyl: vinyl, turntableConfiguration: turntableConfiguration, delegate: delegate, delegateQueue: delegateQueue)
     }
     
-    public convenience init(vinylName: String, bundle: NSBundle = testingBundle(), turntableConfiguration: TurntableConfiguration = TurntableConfiguration(), delegateQueue: NSOperationQueue? = nil) {
+    public convenience init(vinylName: String, bundle: NSBundle = testingBundle(), turntableConfiguration: TurntableConfiguration = TurntableConfiguration(), delegate: NSURLSessionDelegate? = nil, delegateQueue: NSOperationQueue? = nil) {
         
         let plastic = Turntable.createVinylPlastic(vinylName, bundle: bundle)
-        self.init(vinyl: Vinyl(plastic: plastic), turntableConfiguration: turntableConfiguration, delegateQueue: delegateQueue)
+        self.init(vinyl: Vinyl(plastic: plastic), turntableConfiguration: turntableConfiguration, delegate: delegate, delegateQueue: delegateQueue)
     }
     
     // MARK: - Private methods
 
-    private func playVinyl<URLSessionTask: URLSessionTaskType>(request request: NSURLRequest, fromData bodyData: NSData? = nil, completionHandler: RequestCompletionHandler) throws -> URLSessionTask {
+    private func playVinyl(request request: NSURLRequest, fromData bodyData: NSData? = nil, completionHandler: RequestCompletionHandler) throws -> URLSessionUploadTask {
 
```

Comment:
I can add it back if you'd like. Though I don't think it'll buy you the flexibility you seem to think it will.

The only type that will conform to this going forward is `URLSessionUploadTask`. So even though it uses generic syntax there is nothing generic about its usage (e.g., protocols for protocols sake).

What is more going forward `URLSessionDataTask` and `URLSessionUploadTask`, the only two types that ever conformed to `URLSessionTask`, are unlikely to be able to conform to the protocol as written. So if they should ever need to conform to some protocol in the future you'd be either adding a new protocol or re-writing the existing one (e.g., the same place your at now by just removing it in the first place).

In fact, after thinking through this I think the more prudent solution is to move the current implementation of `playVinyl` to it's _only_ call-site and remove `playVinyl` entirely.

e.g.,

```patch
diff --git a/Vinyl/Turntable.swift b/Vinyl/Turntable.swift
index dce17b7..167884e 100644
--- a/Vinyl/Turntable.swift
+++ b/Vinyl/Turntable.swift
@@ -56,21 +56,6 @@ public final class Turntable: NSURLSession {

     // MARK: - Private methods

-    private func playVinyl(request request: NSURLRequest, fromData bodyData: NSData? = nil, completionHandler: RequestCompletionHandler) throws -> URLSessionUploadTask {
-
-        guard let player = player else {
-            fatalError("Did you forget to load the Vinyl? <U+1F3B6>")
-        }
-
-        let completion = try player.playTrack(forRequest: transformRequest(request, bodyData: bodyData))
-
-        return URLSessionUploadTask {
-            self.operationQueue.addOperationWithBlock {
-                completionHandler(completion.data, completion.response, completion.error)
-            }
-        }
-    }
-
     private func transformRequest(request: NSURLRequest, bodyData: NSData? = nil) -> NSURLRequest {
         guard let bodyData = bodyData else {
             return request
@@ -113,9 +98,18 @@ extension Turntable {
     }

     public override func uploadTaskWithRequest(request: NSURLRequest, fromData bodyData: NSData?, completionHandler: (NSData?, NSURLResponse?, NSError?) -> Void) -> NSURLSessionUploadTask {
+
+        guard let player = player else {
+            fatalError("Did you forget to load the Vinyl? <U+1F3B6>")
+        }

         do {
-            return try playVinyl(request: request, fromData: bodyData, completionHandler: completionHandler) as URLSessionUploadTask
+            let completion = try player.playTrack(forRequest: transformRequest(request, bodyData: bodyData))
+            return URLSessionUploadTask {
+                self.operationQueue.addOperationWithBlock {
+                    completionHandler(completion.data, completion.response, completion.error)
+                }
+            }
         }
         catch Error.TrackNotFound {
             errorHandler.handleTrackNotFound(request, playTracksUniquely: turntableConfiguration.playTracksUniquely)
```